manubot/pandoc/cite.py

```diff
@@ -0,0 +1,177 @@
+from manubot.process.util import (
+    get_citekeys_df,
+    generate_csl_items,
+    load_manual_references,
+)
+
+
+global_variables = {
+    "manuscript_citekeys": list(),
+}
+
+
+def parse_args():
+    """
+    Read command line arguments
+    """
+    parser = argparse.ArgumentParser(
+        description="Pandoc filter for citation by persistent identifier"
+    )
+    parser.add_argument("target_format")
+    parser.add_argument("--pandocversion", help="The pandoc version.")
+    parser.add_argument(
+        "--input", nargs="?", type=argparse.FileType("r", encoding="utf-8"),
+    )
+    parser.add_argument(
+        "--output", nargs="?", type=argparse.FileType("w", encoding="utf-8"),
+    )
+    args = parser.parse_args()
+    return args
+
+
+def _get_citekeys_action(elem, doc):
+    """
+    Panflute action to extract citationId from all Citations in the AST.
+    """
+    if not isinstance(elem, pf.Citation):
+        return None
+    manuscript_citekeys = global_variables["manuscript_citekeys"]
+    manuscript_citekeys.append(elem.id)
+    return None
+
+
+def _citation_to_id_action(elem, doc):
+    """
+    Panflute action to update the citationId of Citations in the AST
+    with their manubot-created keys.
+    """
+    if not isinstance(elem, pf.Citation):
+        return None
+    mapper = global_variables["citekey_shortener"]
+    if elem.id in mapper:
+        elem.id = mapper[elem.id]
+    return None
+
+
+def _get_reference_link_citekey_aliases(elem, doc):
+    """
+    Based on TypeScript implementation by phiresky at
+    https://github.com/phiresky/pandoc-url2cite/blob/b28374a9a037a5ce1747b8567160d8dffd64177e/index.ts#L118-L152
+
+    Uses markdown's link reference syntax to define citekey aliases (tags)
+    https://spec.commonmark.org/0.29/#link-reference-definitions
+    """
+    if type(elem) != pf.Para:
+        return
+    while (
+        len(elem.content) >= 3
+        and type(elem.content[0]) == pf.Cite
+        and len(elem.content[0].citations) == 1  # differs from pandoc-url2cite
+        and type(elem.content[1]) == pf.Str
+        and elem.content[1].text == ":"
+    ):
+        space_index = 3 if type(elem.content[2]) == pf.Space else 2
+        destination = elem.content[space_index]
+        if type(destination) == pf.Str:
+            # paragraph starts with [@something]: something
+            # save info to citekeys and remove from paragraph
+            citekey = elem.content[0].citations[0].id  # differs from pandoc-url2cite
+            citekey_aliases = global_variables["citekey_aliases"]
+            if citekey in citekey_aliases:
+                logging.warning(f"duplicate citekey {citekey}")
+            citekey_aliases[citekey] = destination.text
+            # found citation, add it to citekeys and remove it from document
+            elem.content = elem.content[space_index + 1 :]
+        if len(elem.content) > 0 and type(elem.content[0]) == pf.SoftBreak:
+            elem.content.pop(0)
```

Comment:
@phiresky following our discussing in https://github.com/phiresky/pandoc-url2cite/issues/1, I was curious to see if we could add the link-reference-syntax for defining citekey aliases to our in-progess pandoc filter.

I based this function heavily off of your [`extractCiteKeys`](https://github.com/phiresky/pandoc-url2cite/blob/b28374a9a037a5ce1747b8567160d8dffd64177e/index.ts#L118-L152) function. I noticed your code is [AGPL licensed](https://github.com/phiresky/pandoc-url2cite/blob/master/LICENSE.md). For this repo, we use the BSD 3-Clause License, which is more permissive and hence incompatible with AGPL. Are you okay with us reusing your logic here under this repository's BSD license? We would retain the link in the documentation to your package and elsewhere as appropriate.

Reply:
I hereby release the code in the extractCiteKeys function to the public domain. Feel free to use it in BSD 3-Clause licensed code. I guess I wouldn't have any legal claim since you rewrote it anyways though.

My implementation might have some robustness issues, so don't trust it. It would be best if it was exactly the same as the normal markdown syntax but my implementation was enough for my purposes for now. (Defining link references in random places within paragraphs is ugly anyways). 

This would be another issue that we might want to open in the pandoc repo: the suggestion to somewhat re-unify the syntax for citations and links regarding the handling of both `[@test](http://...)` and `[@test] .... [@test]: http://`. In my opinion the parsing should be as close as possible to normal markdown when citation parsing is active (e.g. to commonmark or to `pandoc --format markdown-citations`). Both for familiarity for new pandoc users as well as for writing "polyglot" pandoc markdown code that "mostly" works with other processors (e.g. github markdown).

Reply:
> Feel free to use it in BSD 3-Clause licensed code. I guess I wouldn't have any legal claim since you rewrote it anyways though.

Thanks! I would say I more translated it than completely rewrote it... so probably a gray area in terms of whether copyright applies.

> In my opinion the parsing should be as close as possible to normal markdown when citation parsing is active

So you think `pandoc --format markdown+citations` should treat `[@test](http://...)` like a hyperlink on `@test` without processing `@test` as a citation? Or process `@test` and also hyperlink it?

Reply:
I think it should still treat it as a citation and either ignore what's written in brackets or somehow try to interpret it as cite data or just add it as metadata to the citation in the AST... or process it but also link it like you said. Not sure. Mainly I think it should not write `(http://...)` as literal text to the output like it currently does since that also doesn't happen in non-pandoc markdown.

Same for `[@test][bla]   \n [bla]: 123` which is `<a href=123>@test</a>` in markdown but `<span class="citation" data-cites="test">[@test]</span><a href="123">bla</a>`.